### ISSUE TITLE
Router and INP (Interactive to Next Paint) metrics

### ISSUE THREAD

**Issue by webberig** [angular/angular#51211]
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

While I was investigating INP metrics on a website built using Angular and Universal, I noticed that browsers are repporting bad INP results because of page navigation. I created a new project with some pages and started noticing peculiar interactions between routing and INP metrics reported.

Source code: https://github.com/webberig/inptest
Live demo: https://webberig-inptest.netlify.app/

The app contains 4 pages and routes:
- Page 1: most basic route setup, small html content
- Page 2: route with a resolver that delays for 5 seconds, small html content
- Page 3: no resolver/delay, fairly large HTML table rendered
- Page 4: resolver that delays for 10ms, the same HTML table as page 3

With dev tools open and the Web vitals extension installed, I noticed some weird results:
- Pages 1 and 3 report INP metrics. Pages 2 and 4 do not
- Page 3 has a bad INP score, page 1 is good

=> I seem to be able to avoid INP metrics by adding an async resolver to the routes.

I don't know if measuring INP during soft-navs is intended or not. Either way, some changes may be needed to the router to optimize (or avoid) INP metrics.

### Please provide a link to a minimal reproduction of the bug

https://webberig-inptest.netlify.app/

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 15.2.9
Node: 18.12.1
Package Manager: npm 8.19.2
OS: darwin x64

Angular: 15.2.9
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1502.9
@angular-devkit/build-angular   15.2.9
@angular-devkit/core            15.2.9
@angular-devkit/schematics      15.2.9
@schematics/angular             15.2.9
rxjs                            7.8.1
typescript                      4.9.5
```

**Comment by JeanMeche** [comment]
Looks like a dupe of #51131 / #42489. Can you confirm ?

**Comment by webberig** [comment]
I did come across those issues, but I doubt having change events and/or displaying placeholders will solve the issue.

Page 2 in my demo has a delay of 5 seconds before the user sees a change, but does not trip INP measurements. This is a false positive.
Page 4 is proof I can mitigate metrics without having to solve the problems I present in page 3.

It's just weird that any delay introduced by a resolver would mitigate INP metrics.

**Comment by matthew-whitfield-uk** [comment]
I think route 2 in your demo with a 5000ms delay should give a 5-second (failed) INP score.  For some reason, it doesn't register correctly in Core Web Vitals. 

Perhaps adding the delay in the canActivate component routing code tricks the browser extension into thinking the interaction has finished after the click. If you move the delay to ngOninit in the component and have it do something complex (to better simulate a component initialising) you will get a 5s failed INP score.

As I put in my issue, from my experience the poor INP scores are from route changes. SPAs, like YouTube and GitHub, put a loading bar at the top of the page at the start of route changes. Showing some immediate initial feedback to the user leads to a good INP score.

The only way I have found to achieve this is by redefining routerLink and wrapping the navigation in a settimeout after showing the loading bar.

**Comment by webberig** [comment]
I did some further experimentation by adding a progress bar. I derived a loading state from the `NavigationStart` and `NavigationEnd` events from the router. [Commit here](https://github.com/webberig/inptest/commit/9d14e57873a757c4102370fb286c47d55bbe6eed)

When navigating the first time from page 1 to page 3, the loading bar appears and I get a good INP score:
```
[Web Vitals Extension] Interaction 24 ms (good)
[Web Vitals Extension] FID 9 ms (good)
[Web Vitals Extension] INP 16 ms (good)
```

When I switch back to page 1 and go to page 3 again, the loading bar no longer appears and the INP has gone bad:
```
[Web Vitals Extension] Interaction 240 ms (needs-improvement)
[Web Vitals Extension] INP 240 ms (needs-improvement)
```

I suspect that the routing is async the first time because it needs to lazyload the Page3 component. This allows the router events observable to draw the loading bar. But the second time however, since the component is already loaded, it looks like navigation is synchronous and blocking up the main thread and stops the progress bar from rendering.

I think this needs to be addressed with our without taking into account the feature request in #42489. Having a loading indicator that is not getting a chance to render will serve no purpose and will not improve INP scores.
